---
title: "Analysis Report 1: The High Level of Similarities Between Skin Bacterial Communities On Left And Right Hands"
author: "Emre Ovet"
date: "October 31, 2018"
output: github_document
bibliography: references.bib
csl: bioinformatics.csl
---

# Introduction

The human body hosts millions of bacterial communities that are beneficial or harmful for their host, with varying abundances and phyla [@costello2009bacterial]. One particular organ of the human body, that is, the tissue, has been recently under the scope of the microbiologists due to the diversity and the abundance of bacterial communities on its surface [@findley2013topographic]. One of the research topics on the bacteria living on the skin surface investigates if it is possible to use skin bacterial communities for forensic investigations. This research was conducted by Fierer et al. and they hypothesized that it is possible. In order to test their hypothesis, they carried out three interrelated studies that combine phylogenetic community analysis with high-throughput pyrosequencing methods [@callahan2016; @fierer2010forensic]. They compared bacterial communities on keyboard keys with those found on the finger surfaces of their owners, examined the similarity of bacterial communities on different objects kept in different temperatures and they matched computer mice with individuals who used them by comparing the bacterial community found on the mice with a bacterial community database created by samples obtained from hundreds of people. After analysing their work, we came up with a new question, that is, are the bacterial communities in left and right hands are similar to each other? We hypothesize that they are different from each other since right-handed people tend to make contact with objects using their right hand and left-handed people tend to use their left hand to make contact with objects in daily life. This will result in having a higher variety and abundance of bacteria in right hands of right-handed people compared to their left hand and a higher variety and abundance of bacterium in left hands of the left-handed people compared to their right hand. However, the population of left-handed people is only %10 of World's population [@hardyck1977left], therefore, the overall variety and abundance of bacteria in right hands should be higher than left hands when we consider the total human population in the World. We believe this is an important topic for research because if we can prove the bacterial communities in left and right hands are different from each other in terms of variety and abundance, this might have further implications on forensic investigation using skin bacterial communities. For instance, by analysing the variety and abundance of bacterial communities on the surface of objects in a crime scene, we can have an idea if the criminal is right or left handed and if the criminal turns out to be left-handed, that'll allow the police to eliminate most of the names on their list for possible suspects. It might also have implications on hygiene and sanitation sector. For example, hand sanitizer companies can develop two different sanitizer gel formulas that are specific to eliminate the pathogenic bacteria found only in left and right hands. This would have a great impact on public health, decrease elementary school absenteeisms caused by infections [@hammond2000effect] and also decrease the infection rates in acute care facilities [@hilburn2003use].

 In order to test our hypothesis, we are gonna use the data from the keyboard experiment of Fierer et al. which can be found at the NCBI sequence read archive with the study number ERP022626. Our primary focus is going to be on left and right shift keys to present the bacterial communities on left and right hands in a miniscale. Using the DADA2 package, we are gonna create an amplicon sequence variant table that records the number of times each exact amplicon sequence variant observed in each sample [@callahan2016]. Then, using the pyloseq package, we are gonna store all related phylogenetic sequencing data as single experiment-level object [@mcmurdie2013]. After that, we are gonna melt all the metadata together using R and make plots on the most abundant phyla in each key using ggplot to see how similar the keys and the fingertips are overall. Next, we are gonna make plots on the most abundant phyla,family and genera in left and right shift keys to see how similar they are in different levels of taxonomy. Finally, to make sure the information we obtained up until this point is not biased ,that is, both of the keys have bacteria from the same individuals, we are gonna make a final plot showing from which individuals the genera are obtained from in each key.


# Methods

## Sample origin and sequencing
We obtained our data from the keyboard analysis conducted by Fierer et al.[@fierer2010forensic] .For the keyboard analysis, keys from 3 personal computer keyboards which belong to 3 healthy individuals with ages varying from 20 to 35 years were swabbed along with the fingertips of their owners by Fierer et al.. 2 of these individuals share the same working space. The keyboards had not been touched for 30 minutes before swabbing and each keyboard swabbed by Fierer et al. between 10 minute intervals. Then, space bar keys from 15 personal or public computer keyboards located in University of Colorado Campus were swabbed. In order to sample skin surfaces and keyboards, pre- moistured and autoclaved cotton-tipped swabs were used. All the swabs were kept at -80 Celsius before DNA extraction. 

 The genomic DNA was extracted by Fierer et al. using MO BIO Powersoil DNA isolation kit [@fierer2010forensic]. Then, 16S rRNA genes were amplified and used in PCR. Furthermore, agarose gel electrophoresis was conducted to visualize the replicate amplicons. Amplicon DNA concentrations were analyzed using Quant-iTPicoGreen dsDNA reagent and kit. Then, a pellet was prepared using cleaned and quantified amplicons for pyrosequencing. Fierer et al. used 454 Life Sciences Genome Sequencer FLX instrument by the Environmental Genomics Core Facility at the University of South Carolina.
 
 For sequence analyses, the criteria established for excluding low quality sequences is that all the sequences should be 200 and 300 base pairs in length, have a quality score higher than 25, don't contain ambigious characters, uncorrectable barcode and should contain primer sequence [@fierer2010forensic].

## Computational

First of all, we started our computational analysis by loading genereal-use packages from
bioconductor and GitHub. Then, we set the base path for our input data files from
keyboard study, sorted the path to ensure samples are in order, extracted sample names and specified the full path for each of the filenames_forward_reads [@callahan2016; @mcmurdie2013]. Next, we plotted the quality profiles of each 20 samples. After checking those quality profiles,
we created a diretory called "filtered_reads_path", placed our filtered files inside the filtered/ subdirectory in it and trimmed the lower quality sequences [@callahan2016; @mcmurdie2013]. Then, we produced a table of read counts before and after trimming. This was followed by building error models from each of the samples and visualizing errors with plots to make sure error models matched our data. After checking the plots, we removed all the replicated sequences to increase DADA2's accuracy [@callahan2016; @fierer2010forensic]. Next, we ran DADA2 to apply the core sample inference algorith to the dereplicated data [@callahan2016; @mcmurdie2013]. After DADA2 ran succesfully, we created a new sequence table with "samples" rows and "sequnce variants" columns. This was followed by an histogram of sequence lenghts to check the distribution of trimmed and denoised sequences. Then, we removed the chimeras from our cleaned reads to obtain just non-chimeric reads. After that, we builded a table of pipeline read counts to see how many sequences remain at each step of the pipeline. Next, we assigned taxonomy to each sequence variant. After assigning taxonomy, we extracted sequences to Fasta and builded a phylogeny using DADA2 [@callahan2016; @mcmurdie2013]. Finally, we read in the phylogeny and construct a Phyloseq object [@fierer2010forensic; @mcmurdie2013]. After completing all these steps, we melted all 3 metadata together, explored the dataset and made plots using ggplot.

# Results

After analyzing our data, we found out that the bacterial communities in left and right shift keys have similar phyla present on them with Actinobacteria being the most abundant phylum, but even though the ranking of the phyla in terms of abundance are same, the actual numbers of abundances in two keys are different from each other. For instance, left shift key has a higher abundance in Actinobacteria, Firmicutes, Fusobacteria and Bacteroidetes, whereas left shift has a higher abundance in Candidatus Saccharibacteria. Moreover, the families in Actinobacteria phylum that are found in left and right shift keys are also similar to each other, with *Propionibacteriaceae* being the most abundant family. Once again, the actual number of abundances in two keys are different from each other left shift key has a significant amount of unidentified families and has two additional families present, *Coriobacteriaceae* and *Micronosporaceae*, but they have very small abundances. Right shift key has also two additional families present on it, *Dermabacteraceae* and *Dietziaceae*, but just like the ones in left shift key, they have very small abundances. Furthermore, the genera in *Propionibacteriaceae* found in left and right shift keys are similar to each other, with *Propionibacterium* being the most abundant one. It's abundance is about two times higher in right shift and also both shift keys have a very small amount of *Microlunatus* genus. In addition, the *Propionibacterium* genus comes from the individuals with host subject identities M2 and M3 in both shift keys. In left shift key, however, the abundances of *Propionibacterium* came from M2 and M3 are close to each other whereas in right shift key, the difference is high. Overall, most of the *Propionibacterium* came from M3 in both shift keys. As a result of all these findings, it is possible to say that the similarity between the left and right shift keys is significantly high in terms of the phyla, families and genera found on them as well as the most abundant members of those groups.


```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load general-use packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")

# this package allows for the easy inclusion of literature citations in our Rmd
# more info here: https://github.com/crsh/citr
# and here:
# http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html
library("citr")

# These are the primary packages well use to clean and analyze the data
# this package needs to be installed from bioconductor -- it's not on CRAN
# see info here: https://benjjneb.github.io/dada2/dada-installation.html
library("dada2")

# This to export a fasta of our final denoised sequence variants
library("seqinr")

# To install this you have to install from GitHub
# See more info here: https://github.com/leffj/mctoolsr
# run this -- install.packages("devtools")
# and then this --
library("mctoolsr")

# And this to visualize our results
# it also needs to be installed from bioconductor
library("phyloseq")
```

```{r extract-sample-and-file-names}
# NOTE: Much of the following follows the DADA2 tutorials available here:
# https://benjjneb.github.io/dada2/tutorial.html
# Accessed October 19, 2018

# set the base path for our input data files
path <- "data/raw_data"

# Sort ensures samples are in order
filenames_forward_reads <- sort(list.files(path, pattern = ".fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq
sample_names <- sapply(strsplit(filenames_forward_reads, "\\."), `[`, 1)

# Specify the full path to each of the filenames_forward_reads
filenames_forward_reads <- file.path(path, filenames_forward_reads)
```

```{r check-quality-plots}
# Plots the quality profiles of all twenty samples
plotQualityProfile(filenames_forward_reads[1:115])
```

We can see from the quality profiles that most reads tend to get pretty bad in quality after around 200 bases. 

```{r filter-reads}
# Place filtered files in filtered/ subdirectory
# note this will fail if the directory doesn't exist
filter_path <- file.path("output", "filtered")
filtered_reads_path <- file.path(filter_path,
                                 paste0(sample_names,
                                        "_filt.fastq.gz"))

# See ?filterAndTrim for details on the parameters
# See here for adjustments for 454 data:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
filtered_output <- filterAndTrim(fwd = filenames_forward_reads,
                                 filt = filtered_reads_path,
                                 maxLen = 300,
                                 maxN = 0, # discard any seqs with Ns
                                 maxEE = 3, # allow w/ up to 3 expected errors
                                 truncQ = 2, # cut off if quality gets this low
                                 rm.phix = TRUE,
                                 compress = TRUE,
                                 multithread = FALSE)
```

```{r filtered-read-counts-table}
# produce nicely-formatted markdown table of read counts
# before/after trimming
kable(filtered_output,
      col.names = c("Reads In",
                    "Reads Out"))
```

```{r learn-errors}
# this build error models from each of the samples
errors_forward_reads <- learnErrors(filtered_reads_path,
                                    multithread = FALSE)
```

```{r visualize-errors-with-plots}
# quick check to see if error models match data
# (black lines match black points) and are generally decresing left to right
plotErrors(errors_forward_reads,
           nominalQ = TRUE)
```

```{r dereplicate-sequences}
# get rid of any duplicated sequences
dereplicated_forward_reads <- derepFastq(filtered_reads_path,
                                         verbose = TRUE)

# Name the derep-class objects by the sample names
names(dereplicated_forward_reads) <- sample_names
```

```{r run-dada}
# parameters adjusted based on recommendations for 454 data here:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
dada_forward_reads <- dada(dereplicated_forward_reads,
                           err = errors_forward_reads,
                           HOMOPOLYMER_GAP_PENALTY = -1, # reduce penalty bc 454
                           BAND_SIZE = 32) # performs local alignments bc indels

# check dada results
dada_forward_reads
```

```{r make-sequence-table}
# produce the 'site by species matrix'
sequence_table <- makeSequenceTable(dada_forward_reads)
```

The output table has `r nrow(sequence_table)` rows (samples) and `r ncol(sequence_table)` columns (sequence variants). Notice how we can embed R code directly in our markdown text.

```{r histogram-of-sequence-lengths}
# Quick check to look at distribution of trimmed and denoised sequences
hist(nchar(getSequences(sequence_table)),
     main = "Histogram of final sequence variant lengths",
     xlab = "Sequence length in bp")
```

```{r remove-chimeras}
# Check for and remove chimeras
sequence_table_nochim <- removeBimeraDenovo(sequence_table,
                                            method = "consensus",
                                            multithread = FALSE,
                                            verbose = TRUE)

# What percent of our reads are non-chimeric?
non_chimeric_reads <- round(sum(sequence_table_nochim) / sum(sequence_table),
                            digits = 4) * 100
```

After removing chimeras, we were left with `r non_chimeric_reads`% of our cleaned reads.

```{r table-of-pipeline-read-counts}
# Build a table showing how many sequences remain at each step of the pipeline
get_n <- function(x) sum(getUniques(x)) # make a quick function
track <- cbind(filtered_output, # already has 2 columns
               sapply(dada_forward_reads, get_n),
               rowSums(sequence_table),
               rowSums(sequence_table_nochim))

# add nice meaningful column names
colnames(track) <- c("Input",
                     "Filtered",
                     "Denoised",
                     "Sequence Table",
                     "Non-chimeric")

# set the proper rownames
rownames(track) <- sample_names

# produce nice markdown table of progress through the pipeline
kable(track)
```

```{r assign-taxonomy}
# assigns taxonomy to each sequence variant based on a supplied training set
# made up of known sequences
taxa <- assignTaxonomy(sequence_table_nochim,
                       "data/training/rdp_train_set_16.fa.gz",
                       multithread = FALSE,
                       tryRC = TRUE) # also check with seq reverse compliments

# show the results of the taxonomy assignment
unname(taxa)
```

```{r extract-sequences-to-fasta}
# we want to export the cleaned, trimmed, filtered, denoised sequence variants
# so that we can build a phylogeny - we'll build the phylogeny outside of R
# but we need the fasta file to do so. We keep the names of each sequence as the
# sequence itself (which is rather confusing), because that's how DADA2 labels
# it's columns (e.g. 'species')
# function taken from https://github.com/benjjneb/dada2/issues/88
export_taxa_table_and_seqs <- function(sequence_table_nochim,
                                       file_seqtab,
                                       file_seqs) {
  seqtab_t <- as.data.frame(t(sequence_table_nochim)) # transpose to data frame
  seqs <- row.names(seqtab_t) # extract rownames
  row.names(seqtab_t) <- seqs # set rownames to sequences
  outlist <- list(data_loaded = seqtab_t)
  mctoolsr::export_taxa_table(outlist, file_seqtab) # write out an OTU table
  seqs <- as.list(seqs)
  seqinr::write.fasta(seqs, row.names(seqtab_t), file_seqs) # write out fasta
}

# actually run the function, with the names of the files we want it to create
# and where to put them
export_taxa_table_and_seqs(sequence_table_nochim,
                           "output/sequence_variants_table.txt",
                           "output/sequence_variants_seqs.fa")
```


```{r read-in-metadata-and-create-phyloseq}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                    "fierer_hand_bacteria_SRA_study_ERP022626_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE,
                          row.names = 6) # sets sample IDs to row names

# read in the phylogeny, which was created from the fasta exported above
# in Geneious by aligning the sequences with MAFFT and then building a
# Maximum-Likelihood tree with RAxML
tree_in <- read_tree("output/sequence_variants_MAFFT_FastTree.newick")

# Construct phyloseq object (straightforward from dada2 outputs)
phyloseq_obj <- phyloseq(otu_table(sequence_table_nochim,
                                   taxa_are_rows = FALSE), # sample-spp matrix
                         sample_data(metadata_in), # metadata for each sample
                         tax_table(taxa), # taxonomy for each sequence variant
                         phy_tree(tree_in)) # phylogeny from sequence variants
```

```{r melt-phyloseq}
melted_phyloseq <- psmelt(phyloseq_obj)
melted_phyloseq$sample_source[
  melted_phyloseq$sample_source == "Left_Shift"] <- "Left_shift"
```
```{r exploring-datasets}
names(melted_phyloseq)
```
```{r The-most-abundant-phylum-in-each-key}
melted_phyloseq %>%
  ggplot(aes(x = sample_source,
             y = Abundance,
             fill = Phylum)) +
  geom_col(position = position_dodge()) +
  theme(axis.text = element_text(angle = 90, hjust = 1)) +
ggtitle("The Most Abundant Phylum in Each Key")
```
**Figure 1**
In figure 1, we can see that in every single key and in fingertips, Actinobacteria is the most abundant phylum with a very high difference compared to the second most abundant phylum, Firmicutes. In "I" and "U" keys, this difference is much smaller. Moreover, there is a negligible amount of unidentified phyla in some of the key bacteria.

```{r Abundance-of-phyla-on-2-left-shift-keys-and-right-shift-keys}
melted_phyloseq %>%
  filter(sample_source %in% c("Left_shift", "Left_Shift", "Right_shift")) %>%
  ggplot(aes(x = sample_source, y = Abundance, fill = Phylum)) +
  geom_col(position = position_dodge()) +
    theme(axis.text = element_text(angle = 90, hjust = 1)) +
  ggtitle("Abundance of Phyla on 2 Left Shift Keys and Right Shift Key")
```
**Figure 2**
In figure 2, we can see that left and right shift keys have similar phyla present on them, with Actinobacteria being the most abundant one in both of them. The only differences between the two would be that right shift has a higher amount of Actinobacteria, Firmicutes, Bacteroidetes and Fusobacteria, whereas left shift has a higher amount of Candidatus Saccharibacteria and also has a negligible amount of unidentified phylum. 

```{r Abundance-of-family-in-actinobacteria-phylum-found-in-left-and-right-shift-keys}
melted_phyloseq %>%
  filter(sample_source %in% c("Left_shift", "Left_Shift", "Right_shift")) %>%
  filter(Phylum %in% c("Actinobacteria")) %>%
  ggplot(aes(x = sample_source, y = Abundance, fill = Family)) +
  geom_col(position = position_dodge()) +
    theme(axis.text = element_text(angle = 90, hjust = 1)) +
  ggtitle("Abundance of family in Actinobacteria phylum found
          in left and right shift keys")
```
**Figure 3**
In figure 3, we can see that left and right shift keys have similar families in Actinobacteria phylum present on them, with *Propionibacteriaceae* being the most abundant one in both of them. Even though the order of the family abundances are the same, every single family present in those two keys have different abundances, except that for *Solirubrobacteraceae*, which seems to be the same in both of them. Furthermore, left shift key has a significant amount of unidentified families and two additional families present, *Coriobacteriaceae* and *Micronosporaceae*, with very small abundances. Moreover, right shift key has also two additional families present, *Dermabacteraceae* and *Dietziaceae*, once again, with very small abundances. 

```{r Abundance-of-genus-in-propionibacteriaceae-family-found-in-shift-keys}
melted_phyloseq %>%
  filter(sample_source %in% c("Left_shift", "Left_Shift", "Right_shift")) %>%
  filter(Family %in% c("Propionibacteriaceae")) %>%
  ggplot(aes(x = sample_source, y = Abundance, fill = Genus)) +
  geom_col(position = position_dodge()) +
  theme(axis.text = element_text(angle = 90, hjust = 1)) +
  ggtitle("Abundance of genus in Propionibacteriaceae family
          found in left and right shift keys")
```
**Figure 4**
In figure 4, we can see that left and right shift keys have similar genera in *Propionibacteriaceae* family present on them. The *propionibacterium* is the most abundant genus in both of them. The abundance of this genus is about two times higher in right shift. Moreover, there is a very small abundance of *Microlunatus* in both keys.

```{r The-host-subject-identities-for-the-Propionibacterium-genus-found-in-shift-keys}
melted_phyloseq %>%
  filter(sample_source %in% c("Left_shift", "Left_Shift", "Right_shift")) %>%
  filter(Genus %in% c("Propionibacterium")) %>%
  ggplot(aes(x = sample_source, y = Abundance, fill = host_subject_id)) +
  geom_col(position = position_dodge()) +
  theme(axis.text = element_text(angle = 90, hjust = 1)) +
  ggtitle("The Host Subject Identities For The Propionibacterium
            Genus Found In Shift Keys")
```
**Figure 5**
In figure 5, we can see that the *Propionebacterium* genus found in left and right shift keys comes from the individuals with host subject identities M2 and M3. In left shift, the abundances of *Propionebacterium* came from M2 and M3 is close to each other whereas in right shift, this difference in abundance is higher. Overall, about 1000 of the *Propionebacterium* comes from the host M3 and about 350 from host M2.

# Discussion



# Sources Cited

@Article{callahan2016,
  title = {DADA2: High-resolution sample inference from Illumina amplicon data},
  author = {Benjamin J Callahan and Paul J McMurdie and Michael J Rosen and Andrew W Han and Amy Jo A Johnson and Susan P Holmes},
  journal = {Nature Methods},
  volume = {13},
  pages = {581-583},
  year = {2016},
  doi = {10.1038/nmeth.3869},
}

@Article{mcmurdie2013,
  author = {Paul J. McMurdie and Susan Holmes},
  journal = {PLoS ONE},
  pages = {e61217},
  title = {phyloseq: An R package for reproducible interactive analysis and graphics of microbiome census data},
  volume = {8},
  number = {4},
  year = {2013},
  url = {http://dx.plos.org/10.1371/journal.pone.0061217},
}

@article{fierer2010forensic,
  title={Forensic identification using skin bacterial communities},
  author={Fierer, Noah and Lauber, Christian L and Zhou, Nick and McDonald, Daniel and Costello, Elizabeth K and Knight, Rob},
  journal={Proceedings of the National Academy of Sciences},
  volume={107},
  number={14},
  pages={6477--6481},
  year={2010},
  publisher={National Acad Sciences}
}

@article{hardyck1977left,
  title={Left-handedness.},
  author={Hardyck, Curtis and Petrinovich, Lewis F},
  journal={Psychological bulletin},
  volume={84},
  number={3},
  pages={385},
  year={1977},
  publisher={American Psychological Association}
}

@article{findley2013topographic,
  title={Topographic diversity of fungal and bacterial communities in human skin},
  author={Findley, Keisha and Oh, Julia and Yang, Joy and Conlan, Sean and Deming, Clayton and Meyer, Jennifer A and Schoenfeld, Deborah and Nomicos, Effie and Park, Morgan and Sequencing, NIH Intramural Sequencing Center Comparative and others},
  journal={Nature},
  volume={498},
  number={7454},
  pages={367},
  year={2013},
  publisher={Nature Publishing Group}
}

@article{hilburn2003use,
  title={Use of alcohol hand sanitizer as an infection control strategy in an acute care facility},
  author={Hilburn, Jessica and Hammond, Brian S and Fendler, Eleanor J and Groziak, Patricia A},
  journal={American journal of infection control},
  volume={31},
  number={2},
  pages={109--116},
  year={2003},
  publisher={Elsevier}
}

@article{hammond2000effect,
  title={Effect of hand sanitizer use on elementary school absenteeism},
  author={Hammond, Brian and Ali, Yusuf and Fendler, Eleanor and Dolan, Michael and Donovan, Sandra},
  journal={American Journal of Infection Control},
  volume={28},
  number={5},
  pages={340--346},
  year={2000},
  publisher={Elsevier}
}

@article{costello2009bacterial,
  title={Bacterial community variation in human body habitats across space and time},
  author={Costello, Elizabeth K and Lauber, Christian L and Hamady, Micah and Fierer, Noah and Gordon, Jeffrey I and Knight, Rob},
  journal={Science},
  volume={326},
  number={5960},
  pages={1694--1697},
  year={2009},
  publisher={American Association for the Advancement of Science}
}
